### PR TITLE
[FLINK-10435][yarn]Client sporadically hangs after Ctrl + C

### PR DESCRIPTION
**Pull request by wangyang0918**
## What is the purpose of the change

This PR is to fix the bug when deploying on yarn in attach mode that flink client hangs after `Ctrl + C`. More information could be found at [FLINK-10435](https://issues.apache.org/jira/browse/FLINK-10435).

## Brief change log

* remove the `yarnClient.close()` in `failSessionDuringDeployment` and use a new yarn client in `DeploymentFailureHook`

## Verifying this change

This change can be verified as follows:

* Manually run a command to start flink cluster on yarn in attach mode. `./bin/flink run -m yarn-cluster examples/streaming/WindowJoin.jar`
* When the application is in ACCEPTED status, send the `Ctrl + C`.
* The flink client exits immediately.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java [hunk 1]
```diff
@@ -1526,16 +1525,22 @@ public YarnDeploymentException(String message, Throwable cause) {
...
+			// A new yarn client need to be created in shutdown hook in order to avoid
+			// the yarn client has been closed by YarnClusterDescriptor.
+			this.yarnClient = YarnClient.createYarnClient();
+			this.yarnClient.init(yarnConfiguration);
 		}
 
 		@Override
 		public void run() {
 			LOG.info("Cancelling deployment from Deployment Failure Hook");
+			this.yarnClient.start();
```
Comment (TisonKun) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java]: `this.` can be omitted. Keep consistence with L1542. So does L1543.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java [hunk 2]
```diff
@@ -1241,7 +1241,6 @@ private void failSessionDuringDeployment(YarnClient yarnClient, YarnClientApplic
 			// call (we don't know if the application has been deployed when the error occured).
 			LOG.debug("Error while killing YARN application", e);
 		}
-		yarnClient.stop();
```
Comment (gyfora) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java]: Why did we remove this? Seems like an unrelated change compared to having a new client in the DeploymentFailureHook

**Comment by wangyang0918** [comment]
> Thanks for opening this pull request @wangyang0918 ! Generally it looks good. I have one question:
> 
> What if `YarnClient#start` or `YarnClient#stop` throws an exception? Do you take them into consideration?

@TisonKun Thanks for your comments.

The shut down hook is a best effort clean up. And yarn client only throws `RuntimeException`. Also when we try to start the yarn client in shut down hook, another yarn client of `YarnClusterDescriptor` is running. So it should be started and stopped successfully.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java (hunk 1 above)
Comment (wangyang0918) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java]: Have updated.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java (hunk 2 above)
Comment (TisonKun) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java]: I can see other caller of `failSessionDuringDeployment` such as L525 and L511. Why do you remove this line. IIUC now we don't stop yarnClient in these two cases.

**Comment by TisonKun** [comment]
> > Thanks for opening this pull request @wangyang0918 ! Generally it looks good. I have one question:
> > What if `YarnClient#start` or `YarnClient#stop` throws an exception? Do you take them into consideration?
> 
> @TisonKun Thanks for your comments.
> 
> The shut down hook is a best effort clean up. And yarn client only throws `RuntimeException`. Also when we try to start the yarn client in shut down hook, another yarn client of `YarnClusterDescriptor` is running. So it should be started and stopped successfully.

make sense. I have one more question above.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java (hunk 2 above)
Comment (wangyang0918) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java]: In L525 and L511, it will throw `YarnDeploymentException` and then yarn client will be stopped in `CliFrontend->runProgram->clusterDescriptor.close()`. We should not stop yarn client twice. 

We create yarn client outside and i think we should always use `clusterDescriptor.close()` to stop the yarn client.

Comment (TisonKun) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java]: Thanks for your clarification. Sounds reasonable.
